Sources/NIOSingletonResources/SingletonType.swift

```diff
@@ -0,0 +1,74 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if canImport(Darwin)
+import Darwin
+#elseif os(Windows)
+import ucrt
+import WinSDK
+#elseif canImport(Glibc)
+import Glibc
+#elseif canImport(Musl)
+import Musl
+#else
+#error("The concurrency NIOLock module was unable to identify your C library.")
+#endif
+import NIOCore
+
+/// Singleton resources provided by SwiftNIO for programs & libraries that don't need full control over all operating system resources.
+///
+/// SwiftNIO allows and encourages the precise management of all operating system resources such as threads and file descriptors.
+/// Certain resources (such as the main `EventLoopGroup`) however are usually globally shared across the program. This means
+/// that many programs have to carry around an `EventLoopGroup` despite the fact the don't require the ability to fully return
+/// all the operating resources which would imply shutting down the `EventLoopGroup`. This type is the global handle for singleton
+/// resources that applications (and some libraries) can use to obtain never-shut-down singleton resources.
+///
+/// Programs and libraries that do not use these singletons will not incur extra resource usage, these resources are lazily initialized on
+/// first use.
+public enum NIOSingletons {}
```

Comment:
Bikeshedding... I don't love the "singletons" in the name. I think "shared" or "global" express the purpose bit more.

Reply:
I wanted to be 100% explicit that this is a singleton so I thought that's better. "Shared"/global is a little overloaded because people usually recommend to "share a global ELG" even if that's not this singleton.

Reply:
Adding to this: Why do we want to introduce new libraries for these singletons instead of adding them to the current libraries? Also from a discoverability standpoint wouldn't it be better for users if they could write something `MTELG.shared` instead of going through `NIOSingletons.multiThreadedPosixEventLoopGroup`

Reply:
SwiftNIO is a low-level high-performance library that wants to give you control. Singletons are the antithesis of that so I really don't want to pollute `NIOCore`/`NIOPosix` with things that really go against the philosophy. I think it's a real feature that you need to let down your pants and say `import NIOSingletons`, that's `git grep`'able so you can easily check if things where the bodies are buried.

Reply:
`shared`/`sharedInstance` are commonly used to refer to singletons, especially in Apple frameworks:
- https://developer.apple.com/documentation/swift/managing-a-shared-resource-using-a-singleton/
- https://developer.apple.com/search/?q=shared&type=Documentation
- https://developer.apple.com/search/?q=sharedInstance&type=Documentation

Reply:
@glbrntt yeah, I know but GRPC/AHC and others use `.shared(...)` for "sharing another ELG" not the global singleton. So I thought it's confusing in NIO land.

I won't fight this but I think NIOSingletons is good because it says what it does. Sure, we can make it `NIOSharedResources` or something but that's hiding the fact that it's a singleton. Maybe we make the module `NIOSingletons*` and the types `NIOSharedResources` or smth?

Reply:
> SwiftNIO is a low-level high-performance library that wants to give you control. Singletons are the antithesis of that so I really don't want to pollute `NIOCore`/`NIOPosix` with things that really go against the philosophy. 

I think it depends on how strongly we want to encourage users to reach for the singletons. If our position is "use them unless you know better" then I don't think we should hide them away in a separate module where they're hard to discover.

I like Franz's suggestion of putting them on the type which could be extended further:

```swift
extension EventLoopGroup where Self == MultiThreadedEventLoopGroup {
    public static var sharedMultiThreadedEventLoopGroup: MultiThreadedEventLoopGroup {
        MultiThreadedEventLoopGroup.shared
    }
}

let bootstrap = ClientBootstrap(group: .sharedMultiThreadedEventLoopGroup)
```

Reply:
So if we stick the extension into `NIOSingletonsPosix` then I think the extension is a good idea.

I like `MTELG.sharedMultiThreadedEventLoopGroup` and `MTELG.sharedSingleThreadedEventLoopGroup` but I think they should be in a singletons module.

Reply:
I think this comes back to discoverability. 

I'm not opposed to a separate module, I just think it needs to be easy for users to find. For the posix singletons we could re-export them from the `NIO` module. AHC and the like should depend on `NIOCore`/`NIOPosix` etc so that would solve that issue. One problem is that NIOTS can't re-export its singletons from NIO...

Reply:
Right, I don't think _users_ should find it.

But let me be more precise:
- I would expect most users (say users of AHC / GRPC / Vapor / ...) to use async/await exclusively. Today, they do indeed need to think/create about ELGs but with this PR I'd hope to get the `import NIO*` totally out of their codebases (yes, it'll take a while until that's fully true)
- I would expect library developers (i.e. people who write AHC / GRPC / Vapor / ...) to know how to use NIO and I'd expect them to be somewhat deliberate about their ELG choices. But today they're in a tough spot: They need to put the ELG stuff into their API and the users need to fill it in. But I'd hope that this PR fixes that. If I made a new library FooClient, I think I would do `actor FooClient { init(eventLoopGroup: ELG = NIOSingletons.multiThreadedPosixGroup) { ... } )`. That way the library developer will need to `import NIOSingletons` in _one file_ and that's it. On the "inside" of their library they'll use `self.group` (and they shouldn't reach for the singletons).

Reply:
I'm inclined to want us to add extensions to the appropriate types, as suggested by Franz and George.

Candidly, I don't think a singletons module is going to help us avoid the problems @weissi describes. It is not really possible to hit the combined goal of a) make it easy for novice users to just use a singleton ELG when it is appropriate and also b) make it really hard for library developers to misuse the singletons. We have tried (and failed) to communicate issues this subtle before, and I'd expect us to fail again. Where users _can_ pull an ELG out of thin air, they _will_.

With that tension in mind I think we get a better result by moving more in one direction than the other. And we're already all the way towards optimizing for avoiding misuse. So if we want to try something else, I'd favour optimizing hard for the novice use-case, and trusting experienced developers to deploy best practices. This is in line with Swift's overall pattern anyway. The cost to expert develoers of messing this up is basically only twofold: harder testing, and lower performance. Neither of these is a game-breaker.

Reply:
But who do you expect to even use the singleton ELGs? Going forward I would not expect to find much NIO code in typical applications, I think they will just use a HTTP client/server, a GRPC client/server, ... All of these can default/have the option to use the singletons without a user ever typing either `import NIO` or `import NIOSingletons`, right?

Reply:
As an immediate piece of feedback I think the 80% answer is "the NIO core team", where the spelling is immaterial. It's not all that easy for the wider ecosystem to adopt these changes in a sprightly fashion. They probably won't. It'll take a long time for this to percolate out.

The broader answer is that library authors will use them, but the majority of library authors are in the long tail. SSWG authors are likely to have their use of ELGs called out in SSWG review. Everyone else in the long tail doesn't get that feedback, and so they'll do whatever seems most "obvious". I worry more about them than about the most plugged-in folks.

But to be clear, I'm happy with these living in a `NIOSingletons` module if you feel strongly. I think it gets extremely ugly extremely fast though, as each new EL type must define its own Singletons module. We end up with NIOSingletons, NIOTSSingletons, NIOWindowsSingletons, etc. I'm not sure that's making it easy to DTRT.

Reply:
I have to side with @weissi in terms of naming - "shared" means too many other things, none of which (as inaccurate as some of those usages may be) would be correct for this.

Reply:
> The broader answer is that library authors will use them, but the majority of library authors are in the long tail. SSWG authors are likely to have their use of ELGs called out in SSWG review. Everyone else in the long tail doesn't get that feedback, and so they'll do whatever seems most "obvious".

Agree with all that. Do you think the most obvious would be to not `import NIOSingletonsPosix` and instead create their own MTELG and leak it?

> But to be clear, I'm happy with these living in a NIOSingletons module if you feel strongly. I think it gets extremely ugly extremely fast though, as each new EL type must define its own Singletons module. We end up with NIOSingletons, NIOTSSingletons, NIOWindowsSingletons, etc. I'm not sure that's making it easy to DTRT.

I don't feel strongly about the particular name `NIOSingletons` vs. `NIOSharedResources` or smth. But I do prefer a separate module. Of course, you correctly point out that we'll end up with one `NIOSingletonsELType` for every EventLoop type. I didn't think that's a problem given that we have 2 today and in the future it'll be 3 or 4 I reckon. But it still is a good point.

If we all feel that having it in `NIO` (more precisely `NIOPosix` but nobody that we're discussing here uses that) then I'm happy to move it over. When I set out to start the PR I was very much not thinking of 'polluting' NIO with singleton badness but I can obviously see where you're coming from.

Reply:
Separate module for the singletons (`import NIOSingletonsPosix` or similar required to get hold of them)

Reply:
Okay, to break the stalemate here, why don't we follow what NIO does already:

- `import NIO` gives you everything -- right for most
- `import NIOCore/Posix/...` gives you only what you need

So `NIO` would `@_exported import NIOSingletonsPosix`.

---

Of course, this still suffers from the issue that `import NIO` potentially gives you access to multiple groups: The "correct" one that's part of your framework as well as the singleton one. Developers will be confused which one to choose and will likely start to default to the singleton ones even in libraries that have a group (like gRPC/AHC/most other libs). IMHO that's bad and unfortunate but most of you disagree so fine.

Reply:
I don't see a big advantage in having a separate target in that case then. For users that just import `NIO` right now they presumably don't care much about what is in our various targets and just want to get it building. Presumably this camp also doesn't have to touch ELGs much or shouldn't once we get them to move over to Concurrency more and more.

On the other hand, where we currently have targeted imports is in most of our libraries and I don't see why we would need to import another target just to get the singleton queues. I would argue that developers who use our targeted imports don't need to import another module just to access the singletons.

Reply:
I don't like it, I think it's the right move (singletons are bad for a reason) but I need the functionality.

So whatever, I put them into `NIOPosix`. Added `MultiThreadedEventLoopGroup.globalSingleton` and `NIOThreadPool.globalSingleton`.